packages/pigeon/lib/cpp_generator.dart

```diff
@@ -757,29 +750,54 @@ void _writeFlutterApiHeader(Indent indent, Api api) {
       indent.writeln('');
       indent.writeln('static const flutter::StandardMessageCodec& GetCodec();');
       for (final Method func in api.methods) {
-        final String returnType = func.returnType.isVoid
-            ? 'void'
-            : _nullSafeCppTypeForDartType(func.returnType);
-        final String callback = 'std::function<void($returnType)>&& callback';
+        final HostDatatype returnType = getHostDatatype(
+            func.returnType,
+            root.classes,
+            root.enums,
+            (TypeDeclaration x) => _baseCppTypeForBuiltinDartType(x));
+        final String successCallback =
+            'std::function<void(${_flutterApiReturnType(returnType)})>&& on_success';
+        const String errorCallback =
+            'std::function<void(const FlutterError&)>&& on_error';
         addDocumentationComments(
             indent, func.documentationComments, _docCommentSpec);
         if (func.arguments.isEmpty) {
-          indent.writeln('void ${func.name}($callback);');
+          indent.writeln('void ${_makeMethodName(func)}($successCallback, '
+              '$errorCallback);');
         } else {
-          final Iterable<String> argTypes = func.arguments
-              .map((NamedType e) => _nullSafeCppTypeForDartType(e.type));
+          final Iterable<String> argTypes = func.arguments.map((NamedType arg) {
+            final HostDatatype hostType = getFieldHostDatatype(
```

Comment:
nit: looks like every call to `getFieldHostDatatype` pulls out the classes and enums field.  It would be cleaner probably just pass the AST if that's the case.